### PR TITLE
Separate recursive solver

### PR DESCRIPTION
**Pull request by zaharidichev**
This PR builds on top of the work that @jackh726 did in #542 and separates the recursive solver into its own crate that depends on chalk_solve. This is just a WIP to get a sanity check whether it is heading in the  right direction. 

One of the questions I have is whether we actually want to limit the API surface are of the recursive solver and chalk engine as much as possible or this is not a goal. 

Namely, is saying that the only thing that the recursive solver crate needs to export is `RecursiveSolverImpl` ?

### CODE REVIEW

File: chalk-engine/Cargo.toml [hunk 1]
```diff
@@ -18,3 +18,7 @@ tracing = "0.1"
 
 chalk-derive = { version = "0.16.0-dev.0", path = "../chalk-derive" }
 chalk-ir = { version = "0.16.0-dev.0", path = "../chalk-ir" }
+chalk-solve = { version = "0.16.0-dev.0", path = "../chalk-solve" }
+
+[dev-dependencies]
+chalk-integration = { path = "../chalk-integration" }
```
Comment (jackh726) [inline chalk-engine/Cargo.toml:24]: Let's add a newline here

File: chalk-engine/src/solve.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+use crate::forest::Forest;
+use crate::slg::aggregate::AggregateOps;
+use crate::slg::{SlgContext, SlgContextOps};
+use chalk_ir::interner::Interner;
+use chalk_ir::{Canonical, ConstrainedSubst, Goal, InEnvironment, UCanonical};
+use chalk_solve::{RustIrDatabase, Solution, Solver, SubstitutionResult};
+
+use std::fmt;
+
+pub struct SLGSolverImpl<I: Interner> {
```
Comment (jackh726) [inline chalk-engine/src/solve.rs:11]: I think we can rename this. Probably just `SLGSolver`

File: chalk-integration/Cargo.toml [hunk 3]
```diff
@@ -14,8 +14,10 @@ string_cache = "0.8.0"
 salsa = "0.10.0"
 tracing = "0.1"
 
+
```
Comment (jackh726) [inline chalk-integration/Cargo.toml:17]: Let's remove this newline

File: chalk-recursive/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+[dependencies]
+rustc-hash = { version = "1.1.0" }
+tracing = "0.1"
+
+chalk-derive = { version = "0.16.0-dev.0", path = "../chalk-derive" }
+chalk-ir = { version = "0.16.0-dev.0", path = "../chalk-ir" }
+chalk-solve = { version = "0.16.0-dev.0", path = "../chalk-solve" }
+
+[dev-dependencies]
+chalk-integration = { path = "../chalk-integration" }
```
Comment (jackh726) [inline chalk-recursive/Cargo.toml:24]: Let's add a newline

File: chalk-recursive/src/recursive.rs [hunk 5]
```diff
@@ -35,11 +32,29 @@ pub(crate) struct RecursiveContext<I: Interner> {
 /// so that each question is answered with effectively a "clean slate"**. This
 /// allows for better caching, and simplifies management of the inference
 /// context.
-pub(crate) struct Solver<'me, I: Interner> {
+struct Solver<'me, I: Interner> {
     program: &'me dyn RustIrDatabase<I>,
     context: &'me mut RecursiveContext<I>,
 }
 
+pub struct RecursiveSolverImpl<I: Interner> {
```
Comment (jackh726) [inline chalk-recursive/src/recursive.rs:40]: Let's also rename this to be `RecursiveSolver`

File: chalk-recursive/src/recursive.rs [hunk 6]
```diff
@@ -277,3 +292,54 @@ impl<'me, I: Interner> SolveDatabase<I> for Solver<'me, I> {
...
+            })
+    }
+
+    fn solve_multiple(
+        &mut self,
+        _program: &dyn RustIrDatabase<I>,
+        _goal: &UCanonical<InEnvironment<Goal<I>>>,
+        _f: impl FnMut(chalk_solve::SubstitutionResult<Canonical<ConstrainedSubst<I>>>, bool) -> bool,
+    ) -> bool {
+        unimplemented!()
```
Comment (jackh726) [inline chalk-recursive/src/recursive.rs:343]: Let's add a reason here: `unimplemented!("Recursive solver doesn't support multiple answers")`

File: chalk-solve/Cargo.toml [hunk 7]
```diff
@@ -27,5 +26,5 @@ chalk-integration = { path = "../chalk-integration" }
 [features]
 default = ["slg-solver", "recursive-solver"]
 
-slg-solver = ["chalk-engine"]
+slg-solver = []
 recursive-solver = []
```
Comment (jackh726) [inline chalk-solve/Cargo.toml:30]: Can we remove these now?

File: chalk-solve/src/infer/normalize_deep.rs [hunk 8]
```diff
@@ -17,7 +17,8 @@ impl<I: Interner> InferenceTable<I> {
     /// See also `InferenceTable::canonicalize`, which -- during real
     /// processing -- is often used to capture the "current state" of
     /// variables.
-    pub(crate) fn normalize_deep<T: Fold<I>>(&mut self, interner: &I, value: &T) -> T::Result {
+    // XXX: can be moved to chalk-engine
+    pub fn normalize_deep<T: Fold<I>>(&mut self, interner: &I, value: &T) -> T::Result {
```
Comment (jackh726) [inline chalk-solve/src/infer/normalize_deep.rs:21]: Let's move this :)

File: chalk-solve/src/solve.rs [hunk 9]
```diff
@@ -276,49 +167,12 @@ impl<I: Interner> Solver<I> {
...
-                                crate::recursive::lib::Guidance::Unknown => {
-                                    crate::solve::Guidance::Unknown
-                                }
-                            })
-                        }
-                    })
-            }
-        }
-    }
+        _should_continue: impl std::ops::Fn() -> bool,
```
Comment (jackh726) [inline chalk-solve/src/solve.rs:174]: Hmm, should probably just name this `should_continue`

File: chalk-solve/src/coherence.rs [hunk 10]
```diff
@@ -71,20 +72,19 @@ impl<I: Interner> SpecializationPriorities<I> {
 #[derive(Copy, Clone, Default, PartialOrd, Ord, PartialEq, Eq, Debug)]
 pub struct SpecializationPriority(usize);
 
-impl<'db, I> CoherenceSolver<'db, I>
+impl<'db, I, S, SC> CoherenceSolver<'db, I, S, SC>
 where
     I: Interner,
+    S: Solver<I>,
```
Comment (nikomatsakis) [inline chalk-solve/src/coherence.rs:78]: Can we possibly make this use a `dyn` trait? I think I'd prefer that to keep the generics under control.

Comment (nikomatsakis) [reply chalk-solve/src/coherence.rs:78]: But I'd be happy to consider all that as follow-up work

File: chalk-integration/src/lib.rs [hunk 11]
```diff
@@ -38,3 +42,104 @@ pub struct TypeKind {
...
+    fn default() -> Self {
+        SolverChoice::slg(10, None)
+    }
+}
+
+#[derive(Debug)]
+pub enum SolverImpl {
+    Slg(SLGSolverImpl<ChalkIr>),
+    Recursive(RecursiveSolverImpl<ChalkIr>),
+}
```
Comment (nikomatsakis) [inline chalk-integration/src/lib.rs:93]: dyn trait?

**Comment by nikomatsakis** [comment]
I'm good to merge when @jackh726 is

File: chalk-solve/Cargo.toml (hunk 7 above)
Comment (Mcat12) [reply chalk-solve/Cargo.toml:30]: Seconding this, from my review of the previous PR: https://github.com/rust-lang/chalk/pull/542#discussion_r446659371

Comment (zaharidichev) [reply chalk-solve/Cargo.toml:30]: @Mcat12 sorry, can you elaborate on this. What do we ideally want to do here

Comment (Mcat12) [reply chalk-solve/Cargo.toml:30]: We should be able to remove the slg and recursive feature flags.

**Comment by zaharidichev** [comment]
Oops seems like  I need to move some tests around, hence the build failure

**Comment by jackh726** [comment]
@zaharidichev I fixed it :) Let me do one more quick review then I'll probably merge this.

File: chalk-solve/src/coherence.rs (hunk 10 above)
Comment (jackh726) [reply chalk-solve/src/coherence.rs:78]: I'll file an issue for this.

**Pull request merged**